htlcswitch/switch.go

```diff
@@ -0,0 +1,770 @@
+			prevSatSent = overallSatSent
+			prevSatRecv = overallSatRecv
+
+		case <-s.quit:
+			return
+		}
+	}
+}
+
+// Start starts all helper goroutines required for the operation of the switch.
+func (s *Switch) Start() error {
+	if !atomic.CompareAndSwapInt32(&s.started, 0, 1) {
+		log.Warn("Htlc Switch already started")
+		return nil
+	}
+
+	log.Infof("Htlc Switch starting")
+
+	s.wg.Add(1)
+	go s.startHandling()
+
+	return nil
+}
+
+// Stop gracefully stops all active helper goroutines, then waits until they've
+// exited.
+func (s *Switch) Stop() error {
+	if !atomic.CompareAndSwapInt32(&s.shutdown, 0, 1) {
+		log.Warn("Htlc Switch already stopped")
+		return nil
+	}
+
+	log.Infof("Htlc Switch shutting down")
+
+	s.mutex.Lock()
+	for _, link := range s.links {
+		delete(s.links, link.ChanID())
+		go link.Stop()
+		log.Infof("Remove channel link with ChannelID(%v)", link.ChanID())
+	}
+	s.mutex.Unlock()
+
+	close(s.quit)
+	s.wg.Wait()
+
+	return nil
+}
+
+// AddLink is used to add and start the newly created channel link and start
+// use it to handle the channel updates.
+func (s *Switch) AddLink(link ChannelLink) error {
+	if err := link.Start(); err != nil {
+		return err
+	}
+
+	s.mutex.Lock()
+	s.links[link.ChanID()] = newBoundedLinkChan(numSlots, link)
+	s.mutex.Unlock()
+
+	log.Infof("Added channel link with ChannelID(%v), bandwidth=%v",
+		link.ChanID(), link.Bandwidth())
+	return nil
+}
+
+// GetLink returns the channel link by its channel point.
+func (s *Switch) GetLink(chanID lnwire.ChannelID) (ChannelLink, error) {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	link, ok := s.links[chanID]
+	if !ok {
+		return nil, ErrChannelLinkNotFound
+	}
+
+	return link, nil
+}
+
+// RemoveLink is used to remove and stop the channel link.
+func (s *Switch) RemoveLink(chanID lnwire.ChannelID) error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	link, ok := s.links[chanID]
+	if !ok {
+		return ErrChannelLinkNotFound
+	}
+
+	delete(s.links, link.ChanID())
+	go link.Stop()
+	log.Infof("Remove channel link with ChannelID(%v)", link.ChanID())
+
+	return nil
+}
+
+// RemoveLinks removes all channel links by given node public key.
+func (s *Switch) RemoveLinks(pubKey []byte) error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	var links []ChannelLink
+	for _, link := range s.links {
+		if bytes.Equal(link.Peer().PubKey(), pubKey) {
+			links = append(links, link)
+		}
+	}
+	if links == nil {
+		return errors.Errorf("unable to locate channel link by"+
+			"public key %v", pubKey)
+	}
+
+	for _, link := range links {
+		delete(s.links, link.ChanID())
+		go link.Stop()
+		log.Infof("Remove channel link with ChannelID(%v)", link.ChanID())
+	}
+
+	return nil
+}
+
+// getLinks is helper function which returns the channel links by hop
+// destination id.
+func (s *Switch) getLinks(destination hopID) ([]ChannelLink, error) {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	var result []ChannelLink
+	for _, link := range s.links {
```

Comment:
Lookup by destination is now `O(L)` whereas formerly in was `O(1)` due to the `node -> links` map that we maintained. A dual level map for restoring the former mapping can be used in  order to simplify some of that prior logic around removing from the middle of the slice.